### ISSUE TITLE
A way to avoid `trigger`

### ISSUE THREAD

**Issue by xingrz** [tbruyelle/RxPermissions#36]
Since passing a `trigger` is somewhat different from Rx's idea, I've found a more beautiful way to achieve it by using RxJava's `compose` method:

```java
private Observable.Transformer<Void, Boolean> ensurePermissions(@NonNull String... permissions) {
    return source -> RxPermissions.getInstance(this).request(source, permissions);
}
```

```java
menuItemClicks(R.id.share)
        .compose(bindToLifecycle())
        .compose(ensurePermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE))
        // ......
```

Please consider this more intuitive style.

**Comment by tbruyelle** [comment]
I'm not sure your way works when a configuration change occurs.

Can you test that : 

Ask for a permission not granted yet with your method, see the permission dialog appear, don't answer but rotate your device to trigger a config change, then answer to the permission dialog. If ok your answer should be take into account. If nothing happen, then it means your way doesn't work.

**Comment by xingrz** [comment]
@tbruyelle tested works, of course.

See also: [Don't break the chain: use RxJava's compose() operator](http://blog.danlew.net/2015/03/02/dont-break-the-chain/)

**Comment by xingrz** [comment]
This may be great:

```
RxView.clicks(myButton)
        .compose(RxPermissions.ensure(this, Manifest.permission.WRITE_EXTERNAL_STORAGE))
        .subscribe();
```

**Comment by tbruyelle** [comment]
Wow thanks dude I'm pleasantly surprised. I was aware of the `compose` method, I use it for lifecycle handling, but I didn't know it could be usefull in that case.

That's a lot better than that ugly method signature `request(trigger, permissions)`.

I will integrate that in the framework.

**Comment by tbruyelle** [comment]
That's a little crazy but you gave me the idea to deprecate all current methods and expose only 2 new methods : 

```java
 public static Observable.Transformer<Object, Boolean> ensure(final Context ctx, final String... permissions) {
        return new Observable.Transformer<Object, Boolean>() {
            @Override
            public Observable<Boolean> call(Observable<Object> o) {
                return RxPermissions.getInstance(ctx).request(o, permissions);
            }
        };
    }

    public static Observable.Transformer<Object, Permission> ensureEach(final Context ctx, final String... permissions) {
        return new Observable.Transformer<Object, Permission>() {
            @Override
            public Observable<Permission> call(Observable<Object> o) {
                return RxPermissions.getInstance(ctx).requestEach(o, permissions);
            }
        };
    }
```

For permissions request that requires immediately trigger (without a button), users would have to do something like : 

```java
Observable.just(null) // Too bad there's no something like Observable.always()
     .compose(RxPermissions.ensure(this, perms))
    .subscribe( granted -> {
});
```

The major benefit of that refactoring is it forces the user to use the library correctly. Indeed currently it's possible to write the following code : 

```java
btn.setOnClickListener(v -> RxPermissions.getInstance(this).request(perm).subscribe( { ... ] )
```

But that code doesn't handle configuration change during permission request. So by exposing only *composable* methods, I make this impossible. What do you think ?

I still have other checks to do but this sounds very promising.

**Comment by xingrz** [comment]
It's ok, cool!

According to the [design guidelines](https://www.google.com/design/spec/patterns/permissions.html#permissions-request-patterns), permissions are requested usually just before they're needed. So immediately trigger is not always needed.

**Comment by tbruyelle** [comment]
> permissions are requested usually just before they're needed

unless they're critical permissions https://www.youtube.com/watch?v=iZqDdvhTZj0

**Comment by tbruyelle** [comment]
Fixed with #38
